File: C++/hackerSimulator.cpp
#include <stdio.h>
#include <string.h>
#include <iostream>
#include <map>
#include <fstream>
#include <bits/stdc++.h>
#include <boost/algorithm/string.hpp>
#include <stdexcept>
#include <stdlib.h>

/*
 * O uso do boost ajuda na hora do split
 * eh preciso instalar a biblioteca libboost-dev
 */
using namespace std;
using namespace boost;

/*
 * Criando diretorios padroes
 * --------------------------
 * Usa mapas para armazenar os subdiretorios e os arquivos
 * Arquivos terao como chave o nome e valor o conteudo
 * Diretorios terao como chave o nome e valor o subdiretorio
 *
 */
struct Diretorio {
	map<string, string> arquivos;
	map<string, Diretorio> subdiretorios;
	Diretorio *pai;
};

struct TuplaDirArq {
	Diretorio dir;
	string arq;
};

// Mapa de ips que apontam para servidores
map<string, Diretorio> servidores;

// Diretorio atual no qual o programa esta
Diretorio _dirAtual;

// Servidor onde o usuario esta
Diretorio _server;

/*
 * Funcao ls, recebe uma struct (o diretorio atual)
 * imprime os diretorios em ordem alfabetica e
 * abaixo os arquivos em ordem alfabetica.
 *
 */
string ls(Diretorio& local) {
	string saida;
	for (auto& x: local.subdiretorios) {
		saida.append(x.first + "/");
		saida.append("\n");
	}
	for (auto& x: local.arquivos) {
		saida.append(x.first);
		saida.append("\n");
	}

	return saida;
}

/*
 * Funcao cat, recebe um diretorio e um nome
 * de arquivo e retorna o conteudo do arquivo.
 *
 */
 string cat(Diretorio& dir, string arquivo) {
 	string saida;
	if (dir.arquivos.count(arquivo))
		saida = dir.arquivos[arquivo] + "\n";
	else
		saida = "cat: " + arquivo + ": Arquivo ou diretorio nao encontrado\n";
	return saida;
}

/*
 * Funcao rm, dado um diretório e um arquivo, remove esse arquivo.
 *
 */
string rm(Diretorio& dir, string arquivo) {
	string saida;
	if (dir.arquivos.count(arquivo)) {
		dir.arquivos.erase(arquivo);
	} else
		saida = "rm: nao foi possivel remover " + arquivo + ": Eh um diretorio ou Arquivo nao encontrado\n";
	return saida;
} 

/*
 * Funcao rmdir, dado um diretorio, remove este.
 *
 */
string rmdir(string dirRemover) {
	string saida;
	Diretorio *dir;
	dir = &_dirAtual;
	if (dir->subdiretorios.count(dirRemover))
		dir->subdiretorios.erase(dirRemover);
	else
		saida = "rm: nao foi possivel remover " + dirRemover + ": Eh um Arquivo ou um diretorio nao encontrado\n";
	return saida;
} // farei mais algumas variacoes se forem necessarias

/*
 * Funcao cp recebe um arquivo a ser copiado e um diretorio detino para este.
 *
 */
string cp(Diretorio& dirOrigem, Diretorio& dirDestino, string arquivo) {
	string saida;
	if (dirOrigem.arquivos.count(arquivo) && !dirDestino.arquivos.count(arquivo))
			dirDestino.arquivos[arquivo] = dirOrigem.arquivos[arquivo];
	else
		saida = "pensar em uma frase para ca";
	return saida;
}

/*
 * muda o diretorio atual para o diretorio passado como paramentro;
 */
string cd(Diretorio& local, string dir) {
	string saida;	
	TuplaDirArq retornaDirAndArq(string dirLoc, Diretorio& dirAtual);
	
	if (dir.compare("..") == 0){
		_dirAtual = *local.pai;
	}

	else {
		try {
			Diretorio dirDestino;
			TuplaDirArq tupla = retornaDirAndArq(dir, local);
			dirDestino = tupla.dir;
			_dirAtual = dirDestino;
											
		} catch (char e) {
			saida = "bash: cd: " + dir + ": Arquivo ou diretório não encontrado";
		}
	}

	return saida;
}

/*
 * Comando que limpa a tela.
 */
void clear() {
	system("clear");
}

/*
 * Comando Help - Informa sobre como fazer uso dos outros comandos do jogo
 */

string help() {
	string saida;
	saida.append("\nCAT -  Comando que serve para a leitura de arquivos \nque estejam no diretório");	
	saida.append("corrente // use: cat  nomeDoMeuArquivo\n\n");
	saida.append("LS - Utilizado para listar o conteúdo do diretório corrente.   // use:   ls\n\n");			saida.append("RM - funciona apenas dentro de um diretório. // use:   rm nomeArquivo\n\n"); 			
	saida.append("RMDIR - Comando usado para apagar um diretóeio // use: <rmdir nomeDiretorio>\n\n");			saida.append("CD - Serve para navegar entre diretórios // use: <cd pasta>  ou  <cd /dirPai/.../pasta>\n\n");
	saida.append("EXIT - Use para deslogar do Servidor // use exit\n\n");
	return saida;
}

/*
 * Recebe o comando, olha se ele existe no universo do jo, caso exista
 * faz o processamento necessario na string para gerar o retorno
 * desejado.
 *
 * Retorno - Uma string com a saida desejada ou uma mensagem de erro.
 */
string escolheComando(string comando, Diretorio& diretorio) {

	string comandoLoc;
	string dirLoc;
	if ((comando.compare("ls") != 0) && (comando.compare("clear") != 0)) {
		vector<string> splited;
		split(splited, comando, is_any_of(" "));
 		comandoLoc = splited[0];
 		
		if (splited.size() > 1) {
			dirLoc = splited[1];
		}
	}

	else {
		comandoLoc = comando;
	}

	string saida;
	Diretorio dirTemp;
	string arqTemp;
	saida = "Comando '" + comandoLoc + "' desconhecido\n";
 	TuplaDirArq retornaDirAndArq(string dirLoc, Diretorio& dirAtual);

	if (comandoLoc.compare("ls") == 0) {
 		if (comando.length() == 2) {
			saida = ls(diretorio);
		}
		else {
			saida = "O comando ls nao possui parametros\n";
		}
	}

	else if (comandoLoc.compare("cat") == 0) {
 		saida = cat(_dirAtual, dirLoc);
	}

	else if (comandoLoc.compare("cd") == 0) {
		saida = cd(diretorio, dirLoc);
	}

	else if (comandoLoc.compare("clear") == 0) {
 		if (comando.length() == 5) {
			clear();
			saida = "";
		}
		else {
			saida = "O comando clear nao possui parametros\n";
		}
	}

	else if (comandoLoc.compare("rm") == 0) {
 		saida = rm(_dirAtual, dirLoc);
	}

	else if (comandoLoc.compare("rmdir") == 0) {
		saida = rmdir(dirLoc);
	}

	else if (comandoLoc.compare("help") == 0) {
	       saida = help();	
	}

	return saida;
 }

/*
 * Recebe uma string onde estara o caminho de onde serah executado um comando
 * esse caminho pode estar no formato xx/yy/oo.txt por exemplo
 * onde o diretorio desejado serah yy e o arquivo oo.txt
 *
 * Retorno - Uma struct semelhante a uma tupla, com o dir e o arquivo desejados, ou
 * apenas o dir desejado, vai depender do comando usado.
 *
 *
 */
TuplaDirArq retornaDirAndArq(string dirLoc, Diretorio& dirAtual) {
	Diretorio *dirReturn;
	TuplaDirArq retorno;
	string arqReturn;
	string comp1("/");
	string comp2("");
	vector<string> spliter(string x);
	
	if (dirLoc.compare(comp1) != 0) {
		vector<string> dirLocSplited = spliter(dirLoc);
		arqReturn = dirLocSplited[dirLocSplited.size() - 1];
		
		if (dirLocSplited[0].compare(comp2) == 0) {
			
			dirAtual = _server;
			for (int i=1 ; i<dirLocSplited.size() ; i++) {
				if (dirAtual.subdiretorios.count(dirLocSplited[i])) {
					dirAtual = (dirAtual.subdiretorios[dirLocSplited[i]]);
				}

				else if (!dirAtual.arquivos.count(dirLocSplited[i])) {
					throw invalid_argument( "Diretorio ou Arquivo nao encontrado" );
				}
			}

			dirReturn = &dirAtual;

		}

		else if (dirLocSplited.size() > 1) {

			for (int i=0 ; i<dirLocSplited.size() ; i++) {
				if (dirAtual.subdiretorios.count(dirLocSplited[i])) {
					dirAtual = dirAtual.subdiretorios[dirLocSplited[i]];
				}

				else if (!dirAtual.arquivos.count(dirLocSplited[i])) { 
					throw invalid_argument( "Diretorio ou Arquivo nao encontrado" );
				}
			}

			dirReturn = &dirAtual;
		}

		else if (dirAtual.subdiretorios.count(dirLocSplited[0])){
			dirReturn = &(dirAtual.subdiretorios[dirLocSplited[0]]);
		}	
				
		else if (!dirAtual.arquivos.count(dirLocSplited[0])) {
			throw invalid_argument( "Diretorio ou Arquivo nao encontrado" );
		}	
	}

	else {
		dirReturn = &_server;
	}

	retorno.arq = arqReturn;
	retorno.dir = *dirReturn;

	return retorno;
}

/*
 * Funcao que separa os nomes de diretorios e arquivos
 * para ajudar na busca, retorna um vetor com os nomes
 * em ordem
 *
 */
vector<string> spliter(string comando) {
	vector<string> saida;

	split(saida, comando, is_any_of("/"));

	return saida;
}

/*
 * Loop principal do sistema
 */
void loopSistema() {
	string comando;
	while (1) {
		cout << "Linux => ";
		getline(cin, comando);

		if (comando.compare("exit") == 0) {
			break;
		}
		else {
			cout << escolheComando(comando, _dirAtual);
		}
	}

}

void esperaEnter () {
	cout << "Pressione Enter para continuar..." << endl;
	string comando;
	getline(cin,comando);
}

int main(void) {

	/*
 	* Criando servidores e seus Diretorios/Arquivos
	 */
	
	// O valoe paiDir armazena o pai do diretorio, para o comando (cd ..) funcionar. 

	// Pc do usuario
	Diretorio pcUsuario, home, stash, misc, bin, sys, lib, log;

	// Dentro de home
	
	misc.pai = &home;
	stash.pai = &home;
	home.subdiretorios["stash"] = stash;
	home.subdiretorios["misc"] = misc;

	// Dentro de log
	log.arquivos["logs.txt"] = "@4455_connection:_from_143.203.86.254";

	// Dentro de bin
	bin.arquivos["x-server.sys"] = "100010001011100011000\n01110001100";
	bin.arquivos["os-config.sys"] = "100010001011111111111\n0111100010001100";
	bin.arquivos["bootcfg.dll"] = "000100010001011100011000\n0111000110000\n1111110";
	bin.arquivos["netcfgx.dll"] = "100111000010001011100011000\n01110111001100";
	
	// Definindo os pais dos diretorios
	home.pai = &pcUsuario;
	log.pai = &pcUsuario;
	bin.pai = &pcUsuario;
	sys.pai = &pcUsuario;
	lib.pai = &pcUsuario;
	
	// Principais
	pcUsuario.subdiretorios["home"] = home;
	pcUsuario.subdiretorios["bin"] = bin;
	pcUsuario.subdiretorios["sys"] = sys;
	pcUsuario.subdiretorios["lib"] = lib;	
	pcUsuario.subdiretorios["log"] = log;	
	pcUsuario.pai = &pcUsuario;
	
	// Definindo server e diretorio iniciais
	_server = pcUsuario;
	_dirAtual = home;
	
	cout << "Inicializando..." << endl;

	esperaEnter();
	cout << "Oi.\n....\nIsso é bem estranho... tanto pra você quanto pra mim.\nEu acho que deveria escrever isso no passado, mesmo que seja difícil admitir que acabou.\nMeu nome é Mark, e se você está lendo isso, provavelmente eu já esteja morto.\n" << endl;
	
	esperaEnter();
	cout << "E-mail:\n             de: 4jud3m3@\n             para: xxxxxxxxxxxxxx\n	Oi estranho... não nos conhecemos, mas se você recebeu esse email é porque eu\njá estou morto. Essa mensagem estava programada para ser enviada pelo servidor apos\nele passar 14 dias sem que eu conseguisse logar. Bom, parece que isso realmente\naconteceu.\n\n	Sei que isso parece maluco, mas você a única pessoa que posso confiar nesse\nmomento, preciso de sua ajuda. Você tá dentro?\n	Se estiver a fim, entra no IP\nxxx.xxx.xx.xx, esse é meu servidor. Pega o último log, pois esse log provavelmente é de\nquem me queria fora da jogada. Mas muito cuidado, não quero que você acabe com eu\n "<< endl;
	
	esperaEnter();
	cout << "Se tiver duvidas sobre os comandos digite <help> Sem as setas" <<endl;
	esperaEnter();

	loopSistema();

	return 1;
}

File: haskell/hackerSimulator.hs
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}


import Data.Aeson
import Data.Text
import qualified Data.ByteString.Lazy as B
import GHC.Generics
import GHC.Read
import Help
import Data.List.Split
import System.Console.ANSI

-- Tupla para representar um arquivo
data Arquivo =
 Arquivo { nomeArq  ::  !String,
 conteudo  ::  !String
 } deriving (Show, Generic, Read)

-- | Type of each JSON entry in record syntax.
data Diretorio =
 Diretorio { nome  :: !String,
 subdiretorios  ::  ![Diretorio],
 arquivos  ::  ![Arquivo]
 } deriving (Show,Generic, Read)

-- Objeto que representa um mensagem para ser passada ao
-- usuario
data Mensagem =
 Mensagem { idm :: !Int,
 mensagem :: !String
 } deriving (Show, Generic, Read)

-- Instances to convert our type to/from JSON.

instance FromJSON Diretorio
instance FromJSON Arquivo
instance FromJSON Mensagem

-- | Location of the local copy, in case you have it,
--   of the JSON file.
jsonFile :: FilePath
jsonFile = "servidores.json"

-- Location of JSON Mensagens
jsonFileM :: FilePath
jsonFileM = "mensagens.json"

-- Json com Diretorios e arquivos
getJSON :: IO B.ByteString
getJSON = B.readFile jsonFile

-- Json com mensagens
getJSONM :: IO B.ByteString
getJSONM = B.readFile jsonFileM

-- Retorna um Either (Semelhante a promisse de JavaScript)
-- A partir da leitura do JSON
retornaEither :: IO (Either String [Diretorio])
retornaEither = (eitherDecode <$> getJSON) :: IO (Either String [Diretorio])

-- Retorna um Either para mensagens
retornaEitherM :: IO (Either String [Mensagem])
retornaEitherM = (eitherDecode <$> getJSONM) :: IO (Either String [Mensagem])

-- Retorna um array de Diretorios
-- Recebe como parametro a promisse e olha se ela foi executada com exito
-- quando executada com exito retorna estah direita ou Right
-- Se estiver em Left, estah com erro
lerJSON :: Either String [Diretorio] -> [Diretorio]
lerJSON entrada = 

 case entrada of
  Right ps -> ps 
  Left err -> []

-- Retorna uma lista de mensagens a partir da leitura do JSON de mensagens
lerJSONM :: Either String [Mensagem] -> [Mensagem]
lerJSONM entrada =
 case entrada of
  Right ps -> ps
  Left err -> []

-- Recebe o Nome de um servidor e a partir da leitura do Json
-- Caso esse servidor exista no sistema, retorna o conteudo desse
-- servidor. Nesse caso funciona para os 5 servidores existentes
-- Caso seja adicionado outros, deve modificar a funcao
-- pois ela itera por um numero especifico de elementos em um array
retornaServidor :: [Diretorio] -> String -> Diretorio
retornaServidor [a, b, c, d, e] nome 
 | nome == "135.110.60.200" = a
 | nome == "112.84.211.124" = b
 | nome == "150.189.56.65" = c
 | nome == "220.99.134.37" = d
 | otherwise = e

-- Recebe um Diretorio e retorna o Array de subdiretorios desse Diretorio
retornaSubdiretorios :: Diretorio -> [Diretorio]
retornaSubdiretorios dir = subdiretorios dir

-- Recebe um Diretorio e retorna o Array de Arquivos dele
retornaArquivos :: Diretorio -> [Arquivo]
retornaArquivos dir = arquivos dir

-- Recebe um array e retorna o numero de elementos que ele possui
len :: [a] -> Int
len [] = 0
len (h:t) = 1 + len t

-- Recebe o nome de um subdiretorio desejado e  o array de
-- subdiretorios do diretorio atual se o subdiretorio 
-- estiver presente no diretorio atual, retorna ele
-- Esse array de Diretorios pode ser obtido pelo retorno da funcao
-- retornaSubdiretorios -- Se nao houver o subdiretorio retorna um diretorio vazio
retornaSubdiretorio :: [Diretorio] -> String -> Int -> Diretorio
retornaSubdiretorio array nomeEsperado indiceAtual
 | indiceAtual < len array && ((nome elemento) == nomeEsperado) = elemento
 | indiceAtual < len array = retornaSubdiretorio array nomeEsperado (indiceAtual + 1)
 | otherwise = Diretorio "" [] []  
 where elemento = array !! indiceAtual

-- Recebe o nome de um arquivo e o array de arquivos do diretorio atual, se o arquivo estiver presente
-- no diretorio, retorna ele. Esse array de arquivos pode ser obtido pelo retorno da funcao
-- retornaArquivos -- Se nao ouver o arquivo retorna um arquivo vazio
retornaArquivo :: [Arquivo] -> String -> Int -> Arquivo
retornaArquivo array nomeEsperado indiceAtual
 | indiceAtual < len array && ((nomeArq elemento) == nomeEsperado) = elemento
 | indiceAtual < len array = retornaArquivo array nomeEsperado (indiceAtual + 1) 
 | otherwise = Arquivo "" ""
 where elemento = array !! indiceAtual

-- Funcao que retorna uma mensagem pelo seu id
-- Parametros: Lista de mensagens e ID da mensagem desejada
-- Retorno: Mensagem escolhida ou uma Mensagem que representa um erro
retornaMensagem :: [Mensagem] -> Int -> Mensagem
retornaMensagem [] idIn = (Mensagem 999 "") -- Se o id nao existir
retornaMensagem (h:t) idIn
 | idm h == idIn = h
 | otherwise = retornaMensagem t idIn

-- Funcao que ler o conteudo de uma mensagem
-- Parametros: Mensagem
-- Retorno: Conteudo Da mensagem
retornaConteudoMensagem :: Mensagem -> String
retornaConteudoMensagem msg
 | idm msg == 999 = "Mensagem Inexiste"
 | otherwise = mensagem msg

-- Funcao que retorna o diretorio atual a partir da lista que guarda o caminho necessario
-- para chegar ate ele.
-- O uso de let nessa funcao foi apenas por questao de organizacao, para facilitar o entendimento
-- ela ira acessar os diretorios recursivamente ateh chegar no ultimo nome de diretorio da lista
-- que serah o que o usuario estarah no momento

retornaDiretorioAtual :: Diretorio -> [Diretorio] -> [String] -> Int -> Int -> Diretorio 
retornaDiretorioAtual dir subdirs nomes 0 num
 
 | num > 1 = do
  let servidor = retornaServidor subdirs (nomes !! 0)
  let subDirsServ = retornaSubdiretorios servidor
  retornaDiretorioAtual servidor subDirsServ nomes 1 (num - 1)
 
 | otherwise = retornaServidor subdirs (nomes !! 0)

retornaDiretorioAtual dir subdirs nomes indice 1 = do
 retornaSubdiretorio subdirs (nomes !! indice) 0
 
retornaDiretorioAtual dir subdirs nomes indice num = do
 let subdir = retornaSubdiretorio subdirs (nomes !! (indice)) 0 
 retornaDiretorioAtual subdir (retornaSubdiretorios subdir) nomes (indice + 1) (num - 1) 

-- Funcao que inverte um boolean
notA :: Bool -> Bool
notA True = False
notA False = True

--
--
-- --------------Funcoes de uso no simulador---------
--
--
-- Funcao que reduz o caminho do diretorio atual
-- Recebe o array de Strings que possui o caminho ateh o
-- diretorio atual e elimina o ultimo elemento, caso ja esteja
-- na raiz, apenas retorna o mesmo caminho da entrada
reduzCaminho :: [String] -> Int -> [String]
reduzCaminho caminho indice
 | (len caminho) == 1 = caminho
 | indice == ((len caminho) - 1) = []
 | otherwise = [caminho !! indice] ++ reduzCaminho caminho (indice + 1)

-- Funcao que aumenta o caminho do diretorio atual
-- Usado no CD, quando acessamos um novo diretirui
aumentaCaminho :: [String] -> String -> [String]
aumentaCaminho caminho dir = caminho ++ [dir]

-- Funcao CD Recebe o nome do diretorio desejado, o caminho do diretorio atual
-- e o diretorio atual, se o diretorio desejavel for alcancavel pelo atual
-- retorna o caminho ate o diretorio atual modificado, caso nao seja um destino
-- valido, retorna o caminho do diretorio atual da mesma forma que recebeu
cd :: String -> Diretorio -> [String] -> [String]
cd "" dir caminho = caminho
cd ".." dir caminho = (reduzCaminho caminho 0)
cd destino dir caminho
 | nome retorno /= "" = aumentaCaminho caminho destino
 | otherwise = ["erro"]-- caminho
 where retorno = (retornaSubdiretorio (retornaSubdiretorios dir) destino 0)

-- Recebe uma lista e um elemento e retorna True caso o elemento nao esteja na
-- lista ou False caso ele esteja. Usada na funcao RM
listaNaoPossui :: (Eq a) => [a] -> a -> Bool
listaNaoPossui [] a = True
listaNaoPossui (h:t) a
 | a == h = False
 | otherwise = listaNaoPossui t a

-- Funcao que verifica se um arquivo foi apagado, a partir do nome do arquivo
-- Uma lista de arquivos ja apagados e o diretorio atual. Retorna True
-- Caso o arquivo nao tenha sido apagado e False caso contrario
verificaNomeArquivo :: String -> [String] -> [(String, String)] -> Bool
verificaNomeArquivo nome dirAtual apagados
 | listaNaoPossui apagados (servidor, nome) = True
 | otherwise = False
 where servidor = (dirAtual !! 0) 

-- Funcao que retorna o nome dos subdiretorios de um local
-- usada para a funcao ls
retornaNomesDirs :: [Diretorio] -> Int -> [String]
retornaNomesDirs subs indice 
 | len subs == 0 = []
 | indice == ((len subs) - 1) = [nome (subs !! indice)]
 | otherwise = [nome (subs !! indice)] ++ retornaNomesDirs subs (indice + 1)

-- Funcao qoe retorna os nomes dos arquivos de um local, usada
-- na funcao ls
retornaNomesArqs :: [Arquivo] -> Int -> [String]
retornaNomesArqs arqs indice 
 | len arqs == 0 = []
 | indice == ((len arqs) - 1) = [nomeArq (arqs !! indice)]
 | otherwise = [nomeArq (arqs !! indice)] ++ retornaNomesArqs arqs (indice + 1)

-- Funcao que retorna o menor elemento de uma lista
-- usada no insertion sort
menorLista :: (Ord l) => [l] -> l
menorLista [x] = x
menorLista (h:t) 
 | h <= menorLista t = h
 | otherwise = menorLista t

-- Funcao que remove um elemento de uma lista, usada no insertion sort
removeDaLista :: (Ord l) => [l] -> l -> [l]
removeDaLista [] elem= []
removeDaLista (h:t) elem
 | h == elem = t
 |otherwise = [h] ++ removeDaLista t elem 

-- Funcao que ordena uma lista por meio de um insertion sort
-- pode usar para qualquer lista de elementos ordenaveis
ordenaLista :: (Ord a) => [a] -> [a]
ordenaLista [] = []
ordenaLista lista = [menor]  ++ ordenaLista (removeDaLista lista menor)
 where menor = menorLista lista

-- Funcao que organixa o print da funcao ls, dando espacp entre os nomes e
-- pulando linhas
retornaSaidaLs :: [String] -> Int -> [String] -> [(String, String)] -> String
retornaSaidaLs [] cont dirAtual apagados= ""
retornaSaidaLs (h:t) cont dirAtual apagados
 | cont < 8  && (verificaNomeArquivo h dirAtual apagados) = h ++ " " ++ retornaSaidaLs t (cont + 1) dirAtual apagados
 | cont == 8 = "\n" ++ retornaSaidaLs t 0 dirAtual apagados
 | otherwise = ""

-- Funcao ls - Recebe um dirtorio e retorna seus subdiretorios e seus arquivos 
-- em ordem alfabetica
ls :: Diretorio -> [String] -> [(String, String)] -> String
ls dir dirAtual apagados = retornaSaidaLs (ordenaLista ( (retornaNomesArqs (retornaArquivos dir) 0) ++ ( retornaNomesDirs (retornaSubdiretorios dir) 0) )) 0 dirAtual apagados

-- Funcao connect - Recebe um ip e caso esse ip exista no jogo retorna uma nova 
-- lista para representar o diretorio atual, caso nao exista o ip, retorna uma
-- lista vazia
connect :: String -> [String]
connect "135.110.60.200" = ["135.110.60.200", "home"]
connect "112.84.211.124" = ["112.84.211.124", "home"]
connect "150.189.56.65" = ["150.189.56.65", "home"]
connect "220.99.134.37" = ["220.99.134.37"]
connect outro = ["erro"]

-- Funcao usada para se desconectar de um servidor remoto
disconnect :: [String] -> String
disconnect lista
 | servidor == "135.110.60.200" = "Você não está conectado a um servidor remoto."
 | otherwise = "Desconectando..."
 where servidor = (lista !! 0)

-- Funcao CAT, recebe um diretorio e o nome de um arquivo, se o arquivo existir
-- retorna seu conteudo, caso contrario, retorna uma mensagem informando que
-- o arquivo nao existe
cat :: Diretorio -> String ->[String] -> [(String,String)] -> String
cat dir "" dirAtual apagados = "Informe o nome de um arquivo para ler."
cat dir nome dirAtual apagados
 | nomeArq arquivo == "" || (notA (listaNaoPossui apagados (servidor, nome))) = "Arquivo " ++ nome ++ " nao encontrado."
 | otherwise = conteudo arquivo
 where 
  arquivo = retornaArquivo (retornaArquivos dir) nome 0
  servidor = dirAtual !! 0

-- Funcao que apaga um arquivo de um Diretorio - Parametros : diretorioAtual nomeArquivo
-- arquivosApagados CaminhoDirAtual - Retorna uma nova lista de diretorios apagados
rm :: Diretorio -> String -> [(String, String)] -> [String] -> [(String,String)]
rm dir "" apagados  dirAtual = apagados
rm dir nomeArquivo apagados dirAtual
 | nomeArq arquivo /= "" && (listaNaoPossui apagados (servidor,nomeArquivo)) = apagados ++ [(servidor,nomeArquivo)]
 | otherwise = [("erro","")]
 where 
  arquivo = (retornaArquivo (retornaArquivos dir) nomeArquivo 0)
  servidor = (dirAtual !! 0)

-- Comando que aciona a policia, caso exista provas de um crime contra o proprietario
-- do IP informado.
sshinterpol :: String -> [String] -> [(String,String)] -> String
sshinterpol ip dirAtual apagados
 | ip == "220.99.134.37" && (dirAtual !! 0) == "150.189.56.65" && (listaNaoPossui apagados ("150.189.56.65", "sshinterpol")) = ""
 | (dirAtual !! 0) /= "150.189.56.65" || (not (listaNaoPossui apagados ("150.189.56.65","sshinterpol"))) = "Comando desconhecido: sshinterpol"
 | ip == "" = "Você precisa informar um IP como parâmetro."
 | listaNaoPossui ["220.99.134.37", "150.189.56.65", "112.84.211.124", "135.110.60.200", "localhost"] ip = "HostName não encontrado: " ++ ip
 | otherwise = "Você não pode denunciar alguém sem provas!"

-- Funcao que formata o caminho atual do usuario para exibir no console
-- Parametros: CaminhoDirAtual
-- Saida: CamihoDirAtual formatado
formataCaminhoAtual :: [String] -> String
formataCaminhoAtual [] = ""
formataCaminhoAtual (h:t) = h ++ "/" ++ (formataCaminhoAtual t)

-- Funcao que recebe a entrada do usuario, identifica qual o comando desejado e seus atributos
-- Parametros: Funcao DiretorioAtual Parametro CamihoDirAtual ArquivosApagados
-- Retorno: Saida da funcao desejada
chamaFuncao :: String -> Diretorio -> String -> [String] -> [(String,String)] -> String -> String
chamaFuncao funcao diretorio arquivo dirAtual arquivosApagados mensagem
 | funcao == "" = ""
 | funcao == "clear" = ""
 | funcao == "ls" = ls diretorio dirAtual arquivosApagados
 | funcao == "cat" = cat diretorio arquivo dirAtual arquivosApagados
 | funcao == "getmessage" = mensagem
 | funcao == "sshinterpol" = sshinterpol arquivo dirAtual arquivosApagados
 | funcao == "connect" = do
  let resultadoConnect = connect arquivo
  if (Prelude.head resultadoConnect == "erro") then "Informe um Host válido." else "Connecting to " ++ Prelude.head (connect arquivo) ++ "..."
 | funcao == "cd" = do
  let resultadoCd = cd arquivo diretorio dirAtual
  if (Prelude.head resultadoCd == "erro") then ("Diretorio não encontrado: " ++ arquivo) else ""
 | funcao == "rm" = do
  let resultadoRm = rm diretorio arquivo arquivosApagados dirAtual
  if (resultadoRm == [("erro", "")]) then "Digite um nome de arquivo valido apos o RM." else if (resultadoRm == arquivosApagados) then "Informe o nome de um arquivo para apagar." else ""
 | funcao == "help" = help
 | funcao == "exit" = "Terminando programa..."
 | funcao == "disconnect" = do
  if (arquivo /= "") then ("A funcao Disconnect nao precisa de parametros.") else (disconnect dirAtual)
 | otherwise = "Comando desconhecido: " ++ funcao

retornaNovaMensagem :: String -> Int -> [(String, String)] -> Int
retornaNovaMensagem "connect 112.84.211.124" 4 apagados = 5
retornaNovaMensagem "disconnect" 5 apagados = 6
retornaNovaMensagem "connect 150.189.56.65" 6 apagados = 7
retornaNovaMensagem "sshinterpol 220.99.134.37" 7 apagados = do
 if (listaNaoPossui apagados ("150.189.56.65", "sshinterpol")) then 8 else 7
retornaNovaMensagem "disconnect" 8 apagados = do
 if ((listaNaoPossui apagados ("112.84.211.124", "log.txt")) || (listaNaoPossui apagados ("150.189.56.65", "log.txt"))) then 10 else 9
retornaNovaMensagem entrada idMens apagados= idMens

-- Espera digitar um Enter para mudar a mensagem
esperaEnter :: IO ()
esperaEnter = do 
 putStr "\nDigite Enter para continuar..."
 getLine
 putStr ""

-- Loop principal; recebe um comando, executa ele e depois chama a si mesma com um
-- novo comando.
mainLoop dirAtual arquivosApagados idMsg jaImprimiuMsg = do
  d <- retornaEither
  m <- retornaEitherM
  
  let dirVazio = Diretorio "" [] []
  let diretorio = (retornaDiretorioAtual dirVazio (lerJSON d) dirAtual 0 (len dirAtual))
  
  let mensagemRecebida = retornaConteudoMensagem (retornaMensagem (lerJSONM m) idMsg)
  
  if (not jaImprimiuMsg) then putStrLn mensagemRecebida else (putStr "")
  
  putStr ("root@" ++ (formataCaminhoAtual dirAtual) ++ ":>> ")
  
  entrada <- getLine
  
  let idMensagem = retornaNovaMensagem entrada idMsg arquivosApagados 
  let splitted = Data.List.Split.splitOn " " entrada
  let nomeFuncao = Prelude.head splitted
  let nomeArquivo = if (len splitted > 1) then Prelude.head (Prelude.tail splitted) else ""
  
  let resultChamaFuncao = (chamaFuncao nomeFuncao diretorio nomeArquivo dirAtual arquivosApagados mensagemRecebida)
  if (resultChamaFuncao == "") then (putStr "") else (putStrLn resultChamaFuncao)  
  
  let resultadoTroca = (if nomeFuncao == "cd" then (cd nomeArquivo diretorio dirAtual) else (if nomeFuncao == "connect" then (connect nomeArquivo) else (if (nomeFuncao == "disconnect") then (if (dirAtual !! 0 /= "135.110.60.200") then ["135.110.60.200", "home"] else (dirAtual) ) else dirAtual)))
 
  let novoDirAtual = if (Prelude.head resultadoTroca == "erro") then dirAtual else resultadoTroca
  
  let saidaRm = if nomeFuncao == "rm" then rm diretorio nomeArquivo arquivosApagados dirAtual else arquivosApagados
  let novosArquivosApagados = if saidaRm /= [("erro", "")] then saidaRm else arquivosApagados
  
  if nomeFuncao == "clear" then clearScreen else (putStr "")
  
  let novoJaImprimiuMsg = idMensagem == idMsg
  if entrada == "exit" then (putStr "") else mainLoop novoDirAtual novosArquivosApagados idMensagem novoJaImprimiuMsg -- cada comando aumenta a pilha de recursão! tadinho do stack

main :: IO ()
main = do
  
 d <- retornaEither
 m <- retornaEitherM
 
 let dirAtual = ["135.110.60.200", "home"]
 let arquivosApagados = [("135.110.60.200", "i"),("","")]

 putStr (retornaConteudoMensagem (retornaMensagem (lerJSONM m) 0))
 esperaEnter
  
 putStr (retornaConteudoMensagem (retornaMensagem (lerJSONM m) 1))
 esperaEnter

 putStr (retornaConteudoMensagem (retornaMensagem (lerJSONM m) 2))
 esperaEnter
 
 putStr (retornaConteudoMensagem (retornaMensagem (lerJSONM m) 3))
 esperaEnter

 putStr (retornaConteudoMensagem (retornaMensagem (lerJSONM m) 4))
 esperaEnter

 mainLoop dirAtual arquivosApagados 4 True
